controllers/complianceeventsapi/auth.go

```diff
@@ -0,0 +1,194 @@
+// Copyright Contributors to the Open Cluster Management project
+package complianceeventsapi
+
+import (
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"strings"
+
+	authzv1 "k8s.io/api/authorization/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiserverx509 "k8s.io/apiserver/pkg/authentication/request/x509"
+	"k8s.io/apiserver/pkg/server/dynamiccertificates"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+// parseToken will return the token string in the Authorization header.
+func parseToken(req *http.Request) string {
+	return strings.TrimSpace(strings.TrimPrefix(req.Header.Get("Authorization"), "Bearer"))
+}
+
+// getClientFromToken will generate a Kubernetes client using the input config and token. No authentication data from
+// the input config is used in the generated client.
+func getClientFromToken(cfg *rest.Config, token string) (*kubernetes.Clientset, error) {
+	userConfig := &rest.Config{
+		Host:    cfg.Host,
+		APIPath: cfg.APIPath,
+		TLSClientConfig: rest.TLSClientConfig{
+			CAFile:     cfg.TLSClientConfig.CAFile,
+			CAData:     cfg.TLSClientConfig.CAData,
+			ServerName: cfg.TLSClientConfig.ServerName,
+			Insecure:   cfg.TLSClientConfig.Insecure,
+		},
+		BearerToken: token,
+	}
+
+	return kubernetes.NewForConfig(userConfig)
+}
+
+// canRecordComplianceEvent will perform certificate or token authentication and perform a subject access review to
+// ensure the input user has patch access to patch the policy status in the managed cluster namespace. An error is
+// returned if the authorization could not be determined. Note that authenticatedClient and authenticator can be nil
+// if certificate authentication isn't used. If both certificate and token authentication is present, certificate takes
+// precedence.
+func canRecordComplianceEvent(
+	cfg *rest.Config,
+	authenticatedClient *kubernetes.Clientset,
+	authenticator *apiserverx509.Authenticator,
+	clusterName string,
+	req *http.Request,
+) (bool, error) {
+	postRules := authzv1.ResourceAttributes{
+		Group:       "policy.open-cluster-management.io",
+		Version:     "v1",
+		Resource:    "policies",
+		Verb:        "patch",
+		Namespace:   clusterName,
+		Subresource: "status",
+	}
+
+	// req.TLS.PeerCertificates will be empty if certificate authentication is not enabled (e.g. endpoint is not HTTPS)
+	if req.TLS != nil && len(req.TLS.PeerCertificates) > 0 {
```

Comment:
If not HTTPS will an error be returned?

Reply:
If not HTTPS then `req.TLS` will be `nil` so then only token based authentication is supported.